README
DSSL library: network capture and SSL decryption toolkit

Copyright (C) 2007-2009 Atomic Labs, Inc.

To view the installation documentation, load docs/index.html into your browser.

For UNIX systems:

# ./autogen.sh  (only required if using subversion checkout)
# ./configure   (run ./configure --help to see available options)
# make all      (this will build everything except DSSL's unit tests)
# make install  (this may require superuser/Administrator privileges)

For Universal binaries on OSX, change this line in configure.in:

CFLAGS="-DDSSL"

to this:

CFLAGS="-DDSSL -arch i386 -arch x86_64 -arch ppc -arch ppc64"
 
And run "./configure --disable-dependency-tracking"


ChangeLog
2011-04-14: version 2.1.1
	* Improved reassembly logic for retransmitted/overlapping packets
2011-04-04: version 2.0.4
	* Added new statistics to CapEnvGetMemoryUsage()
	* Fixed problems with infinitely growing server list for SSL traffic
	* Added max tcp sessions and max cached packets constraints
2010-01-18: version 2.0.1
	* Fixed problem with missing callback being called with huge byte counters
	* Process all unacknowledged data when TCP session closes
	* Handle ACK's in RST and other finishing packets
	* Fixed TCP reassembly bug causing wrong packet processing order
	* Fix for TCP reassembly "deadlock" caused by ACK packet sequence
	* Cleaned-up duplicate SessionToString and AddressToString functions
	* Added TLS session ticket table cleanup code
	* Added "link_type" field to DSSL packet structure
	* Added SSL key auto-mapping functionality
	* Fixed SSL v2 handshake protocol parsing regression
	* Fixed SSL v2 session renegotiation parsing
	* Got unit tests working on most Unix platforms
	* Various project file updates
2009-12-29: version 1.7.3
	* Added TCP session timeout parameter to CapEnvCreate
	* Fixed incorrect reassembly of TCP sessions with missing SYN packets
	* Added DSSL_TCPSTREAM_SENT_RST TcpStream flag
2009-10-07: version 1.7.2
	* Added DSSL_EnvSetServerInfoWithKey() function for existing RSA private keys
2009-09-23: version 1.7.1
	* Added memory usage reporting API - see CapEnvGetMemoryUsage for details
	* Fixed max SSL record length check to allow zero-length SSL records
2009-09-15: version 1.7.0
	* Fixed TCP reassembly problem causing SSL session losses if SYN packet from the server is processed before SYN+ACK one 
	* Added max SSL record length check
	* Fixed bug in reassembly queue size check logic
2009-09-07: version 1.6.9
	* Fixed bug in TCP reassembly queue size management
2009-08-14: version 1.6.8
	* Last cleanup time of SSL session key cache was never being updated
	* Fixed severe performance problems with SSL session key cache
2009-08-13: version 1.6.7
	* Added SYN/SYN+ACK and first ACK timestamp tracking in TcpStream
2009-06-20: version 1.6.6
	* Copyright changed to Atomic Labs, Inc.
	* Misc autoconf cleanup to sync with Pion repositories
2009-04-28: version 1.6.5
	* Minor fixes in VC++ project files, unit tests, etc.
2009-03-31: version 1.6.4
	* Fixed spurious missing packet callbacks
	* Now flushing packets remaining in TCP sessions when closing (r79)
2009-03-27: version 1.6.3
        * Reverting flushing logic change (r79) due to regression
	* Added VC project file patch for Windows server 2008
2009-03-26: version 1.6.2
	* Fixed duplicate SYN packet bug
	* Fixed missing SYN,ACK packet bug
	* Improved TCP session flushing logic
2009-03-16: version 1.6.0
	* Fixed multiple session between the same srcIP:port<->targetIP:port
	* Fixed FIN packet handling 
2009-02-09: version 1.5.2
	* Added TCP session statistics
	* Added SSL handshake completed event callback
2009-01-22: version 1.5.1
	* Added missing packet callback capability to support partially captured TCP sessions
	* Added ACK time information to TCP packets
	* Main callback routine changed to include a pointer to the last captured frame 
2009-01-09: version 1.4.6
	* SSL Session ID length check bug fixed - thanks to Michael Dickey from Atomic Labs (http://www.atomiclabs.com) 
	* Fixed TCP packet retransmission reassembly bug
	* Improved out-of-order packet handling performance in TCP reassembler
	* Improved TCP reassembly algorithm for out-of-order TCP packet processing
2008-12-05: version 1.4.4
	* Fixed TCP session hashing bug - thanks to Faisal Shaikh
	* TCP session timeout set to 180 seconds
2008-11-30: version 1.4.3
	* Fixed memory leak in TCP session processing
	* Added TCP session table cleanup
2008-09-25: version 1.4
	* Added ACK control to TCP stream processing
2008-09-18: version 1.3.3
	* Added session start timestamp to the TCP session object
2008-07-11: version 1.3.2
	* Missing const keyword added to SessionSetUserData function
	* TCP session capture can no longer start from the middle of TCP session
2008-07-05: version 1.3.1
	* Fixed wrong ASSERT statement bug in decoder.c
2008-05-06: version 1.3.0
	* Added SSL version 2.0 support
	* Fixed TCP reassembly bug
2008-03-23: version 1.2.0
	* Added DEFLATE compression support (RFC 3749)
	* Added MacOS support - thanks to Michael Dickey from Atomic Labs (http://www.atomiclabs.com)
2007-06-05: version 1.0.3
	* Fixed session management bug causing non-existing empty sessions being created and destroyed
	* Added ssltrace sample - a simple command line utility that dumps decrypted SSL data
	* Added server's certificate check to make sure the cert is signed with the same private key DSSL initialized with
	* New error codes added (see errors.h)
	* Fixed SSL decryption engine: block cipher padding length check added to prevent possible buffer overrun condition
	* Fixed NULL password parameter was causing a segmentation fault (access violation on Win32)
2007-04-19: build 1.0.2
	* DSSL_NO_PCAP define added. Now you can compile DSSL without referencing functions from pcap library.

2007-03-26: version 1.0.1
	* Fix RSA private key file loading errors.
	* Fix memory leak in MAC calculation.

2007-01-17: version 1.0.0
	* First public release.



src/errors.h
/*
** This file is a part of DSSL library.
**
** Copyright (C) 2005-2009, Atomic Labs, Inc.
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 2 of the License, or
** (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
**
*/
#ifndef __DSSL_ERRORS_H__
#define __DSSL_ERRORS_H__

#define NM_IS_FAILED( rc ) ((rc) < 0) 

#define DSSL_RC_WOULD_BLOCK						1
#define DSSL_RC_OK								0
#define DSSL_E_OUT_OF_MEMORY					(-1)
//#define DSSL_E_SSL_LOAD_CERTIFICATE				(-3)
#define DSSL_E_SSL_LOAD_PRIVATE_KEY				(-4)
#define DSSL_E_SSL_UNKNOWN_VERSION				(-5)
#define DSSL_E_INVALID_PARAMETER				(-6)
#define DSSL_E_SSL_PROTOCOL_ERROR				(-7)
#define DSSL_E_SSL_INVALID_RECORD_LENGTH 		(-8)
#define DSSL_E_UNSPECIFIED_ERROR				(-9)
#define DSSL_E_NOT_IMPL							(-10)
#define DSSL_E_SSL_SERVER_KEY_UNKNOWN			(-11)
/* SSL: An undecryptable ciphersuite has been chosen */
#define DSSL_E_SSL_CANNOT_DECRYPT				(-12)
/* SSL: Invalid Pre Master Secret */
#define DSSL_E_SSL_CORRUPTED_PMS				(-13)
/* SSL: Protocol version rollback detected (SSL attack?) */
#define DSSL_E_SSL_PMS_VERSION_ROLLBACK			(-14)
/* SSL: Generic decrytion error */
#define DSSL_E_SSL_DECRYPTION_ERROR				(-15)
/* SSL Protocol error: Bad Finished message digest */
#define DSSL_E_SSL_BAD_FINISHED_DIGEST			(-16)
/* TCP reassembly error */
#define DSSL_E_TCP_CANT_REASSEMBLE				(-17)
/* Unexpected transmission - session is already closed or aborted */
#define DSSL_E_SSL_UNEXPECTED_TRANSMISSION		(-18)
/* Message Authentification Code check failed */
#define DSSL_E_SSL_INVALID_MAC					(-19)
/* Session renegotiation detected, but the previous 
session data is not cached - decryption impossible */
#define DSSL_E_SSL_SESSION_NOT_IN_CACHE			(-20)
/* Failed to open the private key file */
#define DSSL_E_SSL_PRIVATE_KEY_FILE_OPEN		(-21)
/* SSL Protocol error: invalid certificate list length */
#define DSSL_E_SSL_INVALID_CERTIFICATE_RECORD	(-22)
/* SSL Protocol error: invalid certificate length */
#define DSSL_E_SSL_INVALID_CERTIFICATE_LENGTH	(-23)
/* Bad server certificate detected */
#define DSSL_E_SSL_BAD_CERTIFICATE				(-24)
/* Function argument is not initialized properly */
#define DSSL_E_UNINITIALIZED_ARGUMENT			(-25)
/* SSL: ephemeral keys cannot be decrypted */
#define DSSL_E_SSL_CANNOT_DECRYPT_EPHEMERAL		(-26)
/* SSL: Only RSA keys can be used with DSSL */
#define DSSL_E_SSL_CANNOT_DECRYPT_NON_RSA		(-27)
/* SSL: Server's certificate is signed with the key different
than the one passed to CapEnvSetSSL_ServerInfo or 
DSSL_EnvSetServerInfo */
#define DSSL_E_SSL_CERTIFICATE_KEY_MISMATCH		(-28)
#define DSSL_E_UNSUPPORTED_COMPRESSION			(-29)
#define DSSL_E_DECOMPRESSION_ERROR				(-30)
#define DSSL_E_SSL2_INVALID_CERTIFICATE_TYPE    (-31)
#define DSSL_E_SSL2_UNKNOWN_CIPHER_KIND		    (-32)
/* Incorrect SERVER-VERIFY - the data doesn't match the client-sent CHALLENGE */
#define DSSL_E_SSL2_BAD_SERVER_VERIFY		    (-33)
/* Incorrect CLIENT-FINISHED - the data doesn't match the server-sent CONNECTION-ID */
#define DSSL_E_SSL2_BAD_CLIENT_FINISHED		    (-34)
/* Reached the maximum number of out-of-order packts in TCP stream's reassembly queue */
#define DSSL_E_TCP_REASSEMBLY_QUEUE_FULL		(-35)
/* Missing packet detected and user-defined callback code decided to abort the session */
#define DSSL_E_TCP_MISSING_PACKET_DETECTED		(-36)
/* TLS Session Ticket is not cached */
#define DSSL_E_SSL_SESSION_TICKET_NOT_CACHED	(-37)
/* Duplicate SSL server entry */
#define DSSL_E_SSL_DUPLICATE_SERVER				(-38)
/* Duplicate SSL server entry */
#define DSSL_E_TCP_GLOBAL_REASSEMBLY_QUEUE_LIMIT (-39)

#ifdef _DEBUG
	int NmDebugCatchError( int rc );
	#define NM_ERROR( rc ) NmDebugCatchError( rc )
#else
	#define NM_ERROR( rc ) (rc)
#endif

#endif
